### DISCUSSION TITLE
Statically typed dynamic subclass? A case for an intersection type?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#5763] [Q&A] [answered]
One of the repositories I work on ([`flytekit`](https://github.com/flyteorg/flytekit)) has a dependency that depends on [dataclasses-json](https://pypi.org/project/dataclasses-json) which aims to extend the builtin `dataclasses` with some additional json serialization methods. They do this via the [`DataClassJsonMixin`](https://github.com/lidatong/dataclasses-json/blob/89578cb9ebed290e70dba8946bfdb68ff6746755/dataclasses_json/api.py#L18), but `flytekit` never actually used as a mixin. They achieve this "mixing in" via a class decorator [`dataclass_json`](https://github.com/lidatong/dataclasses-json/blob/89578cb9ebed290e70dba8946bfdb68ff6746755/dataclasses_json/api.py#L105) that basically just inserts all the methods of `DataClassJsonMixin` into the wrapped `cls`. This is clearly not static typing friendly, but it occurred to me that a similar implementation would be to just use the mixin as a mixin and create a dynamic subclass using the `dataclass_json` class decorator, but I am having trouble getting pyright to say that the dynamic subclass is both a subclass of the mixin class and the decorated class. My goal was something similar to this:

```python
def f(x):
    assert issubclass(x, int) and issubclass(x, float)
    reveal_type(x)  # Type of "x" is "type[int]* & type[float]*"
```

```python
import abc
import dataclasses
from typing import Optional, Type, TypeVar

T = TypeVar("T")


class DataClassJsonMixin(abc.ABC):
    def to_json(self, **kwargs) -> str:
        ...


def dataclass_json(_cls):
    def wrap(cls):
        return _process_class(cls)
    return wrap(_cls)


def _process_class(cls: Type[T]):  # (cls: Type[T]) -> Type[T] & DataClassJsonMixin ?
    class DynamicDataClassJson(DataClassJsonMixin, cls):
        ...
    
    assert issubclass(DynamicDataClassJson, cls) and issubclass(DynamicDataClassJson, DataClassJsonMixin)
    reveal_type(DynamicDataClassJson)  # Type of "DynamicDataClassJson" is "type[DataClassJsonMixin]*"
    return DynamicDataClassJson


#############################################################################
# pyright doesn't recognize the dynamic subclass as a subclass of the original class

@dataclass_json
@dataclasses.dataclass
class FastSerializationSettings(object):
    enabled: bool = False
    destination_dir: Optional[str] = None
    distribution_location: Optional[str] = None


serialization_settings = FastSerializationSettings(  # Expected no arguments to "DynamicDataClassJson" constructor
    enabled=True,
    destination_dir=".",
    distribution_location="...",
)

print(serialization_settings.to_json())  # OK

reveal_type(serialization_settings)  # Type of "serialization_settings" is "DynamicDataClassJson"

#############################################################################
# this is the desired effect:

@dataclasses.dataclass
class FastSerializationSettings2(object):
    enabled: bool = False
    destination_dir: Optional[str] = None
    distribution_location: Optional[str] = None


class DynamicDataClassJson(DataClassJsonMixin, FastSerializationSettings2):
    ...


serialization_settings = DynamicDataClassJson(  # OK
    enabled=True,
    destination_dir=".",
    distribution_location="...",
)

print(serialization_settings.to_json())  # OK

reveal_type(serialization_settings)  # Type of "serialization_settings" is "DynamicDataClassJson"

```

Is what I am attempting to do feasible in python typing or pyright as it exists today?

**Comment by erictraut** [comment, marked as answer]
Yeah, there's a straightforward way to do this. It's a little more verbose than using the class decorator, but it's not too bad. And it accurately models what's going on at runtime. An intersection would not. By declaring the class statically, a type checker can catch issues like MRO linearization conflicts, metaclass conflicts, member redeclarations, etc.

```python
@dataclass
class _FastSerializationSettings:
    enabled: bool = False
    destination_dir: Optional[str] = None
    distribution_location: Optional[str] = None

class FastSerializationSettings(_FastSerializationSettings, DataClassJsonMixin): ...
```

**Reply by ringohoffman** [reply]
Thanks! This one change eliminates 467 of the 1715 existing pyright errors (27.2%) from [flytekit](https://github.com/flyteorg/flytekit/pull/1801)!
